src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,365 @@
+                if (documentSymbol.Children is not null)
+                    documentSymbolModel.Children = GetDocumentSymbolModels(documentSymbol.Children);
+                documentSymbolModels.Add(documentSymbolModel);
+            }
+
+            return documentSymbolModels.ToImmutableAndFree();
+        }
+
+        public static async Task<JToken?> DocumentSymbolsRequestAsync(
+            ITextBuffer textBuffer,
+            ILanguageServiceBrokerShim languageServiceBroker,
+            string? textViewFilePath,
+            CancellationToken cancellationToken)
+        {
+            if (textViewFilePath is null)
+                return null;
+
+            var parameterFactory = new DocumentSymbolParams()
+            {
+                TextDocument = new TextDocumentIdentifier()
+                {
+                    Uri = new Uri(textViewFilePath)
+                }
+            };
+
+            // TODO: proper workaround such that context.ClientCapabilities?.TextDocument?.DocumentSymbol?.HierarchicalDocumentSymbolSupport == true
+            return await languageServiceBroker.RequestAsync(
+                textBuffer: textBuffer,
+                method: Methods.TextDocumentDocumentSymbolName,
+                capabilitiesFilter: (JToken x) => true,
+                languageServerName: WellKnownLspServerKinds.AlwaysActiveVSLspServer.ToUserVisibleString(),
+                parameterFactory: _ => JToken.FromObject(parameterFactory),
+                cancellationToken: cancellationToken).ConfigureAwait(false);
+        }
+
+        public static void SetIsExpanded(IEnumerable<DocumentSymbolViewModel> documentSymbolModels, bool isExpanded)
+        {
+            foreach (var documentSymbolModel in documentSymbolModels)
+            {
+                documentSymbolModel.IsExpanded = isExpanded;
+                SetIsExpanded(documentSymbolModel.Children, isExpanded);
+            }
+        }
+
+        /// <summary>
+        /// Compares the order of two DocumentSymbolViewModels using their positions in the latest editor snapshot.
+        /// </summary>
+        /// <remarks>
+        /// The parameter oldSnapshot refers to the snapshot used when the LSP document symbol request was made
+        /// to obtain the symbol and currentSnapshot refers to the latest snapshot in the editor. These parameters
+        /// are required to obtain the latest DocumentSymbolViewModel range positions in the new snapshot.
+        /// </remarks>
+        public static int CompareSymbolOrder(DocumentSymbolViewModel x, DocumentSymbolViewModel y, ITextSnapshot oldSnapshot, ITextSnapshot currentSnapshot)
+        {
+            var xStartPosition = oldSnapshot.GetLineFromLineNumber(x.StartPosition.Line).Start.Position + x.StartPosition.Character;
+            var yStartPosition = oldSnapshot.GetLineFromLineNumber(y.StartPosition.Line).Start.Position + y.StartPosition.Character;
```

Comment:
Could we just implement this by comparing the line numbers, and then if those are the same comparing the character numbers? That avoid all the extra translation?